{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "a1487f78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nScientificData论文图片绘制\\n单个基因数据桑吉图\\n'"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "ScientificData论文图片绘制\n",
    "单个基因数据桑吉图\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "df11dacc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "7c23368c",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = '/mnt/disk1/xzyao/CancerPNRLE/merge_data'\n",
    "\n",
    "AML_file = f'{base_path}/Acute Myeloid Leukemia.gene-var.tsv'\n",
    "\n",
    "\n",
    "go_ancestor_file = '/mnt/disk1/xzyao/CancerPNRLE/result/芷涵GO-HPO分级/go_second_ancestors.txt'\n",
    "\n",
    "hpo_ancestor_file = '/mnt/disk1/xzyao/CancerPNRLE/result/芷涵GO-HPO分级/hpo_third_ancestors.txt'\n",
    "\n",
    "\n",
    "fig_save_path = '/mnt/disk1/xzyao/CancerPNRLE/ScientificData投稿富集分析结果/ScientificData_论文绘图/单基因桑吉图绘制'\n",
    "\n",
    "template_html = f'{fig_save_path}/Sankey-template.html'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "0355c1e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def read_obo_level_file(_go_level_file: str):\n",
    "\n",
    "    go_id_to_second_ancestor = {}\n",
    "    with open(_go_level_file) as f:\n",
    "        f.readline()\n",
    "        for line in f:\n",
    "            l = line.strip().split('\\t')\n",
    "            if len(l) != 4:\n",
    "                continue\n",
    "            go_id = l[0]\n",
    "            ancestor_id = l[2]\n",
    "            ancestor_name = l[3]\n",
    "\n",
    "            if ancestor_name.startswith('obsolete'):\n",
    "                ancestor_name = ancestor_name.replace('obsolete', '').strip()\n",
    "\n",
    "            go_id_to_second_ancestor[go_id] = f'{ancestor_id}-{ancestor_name}'\n",
    "\n",
    "    return go_id_to_second_ancestor\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "630bf9e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read ancestor file\n",
    "go_to_ancestor = read_obo_level_file(go_ancestor_file)\n",
    "hp_to_ancestor = read_obo_level_file(hpo_ancestor_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "58bbdf8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def replace_template(template_html: str, nodes: set, links: set, save_file: str):\n",
    "\n",
    "    with open(template_html) as f, open(save_file, 'w') as wf:\n",
    "        doc = f.read()\n",
    "        \n",
    "        doc = doc.replace('data_str', str(nodes))\n",
    "        doc = doc.replace('link_str', str(links))\n",
    "\n",
    "        wf.write(doc)\n",
    "    print(f'{save_file} saved.')\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd4f2a58",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "686a5ae9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_merge_db_file(merge_db_file,\n",
    "                       go_to_ancestor: dict,\n",
    "                       hpo_to_ancestor: dict,\n",
    "                       use_ancestor=False,\n",
    "                       selected_gene='',):\n",
    "    \n",
    "    \n",
    "    node_set = set()\n",
    "\n",
    "    # gene to var_type\n",
    "    gene_var_count = defaultdict(int)\n",
    "    # var_type to sub_var_type\n",
    "    var_sub_var_count = defaultdict(int)\n",
    "    # sub_var_type to bp\n",
    "    sub_var_bp_count = defaultdict(int)\n",
    "    \n",
    "       \n",
    "    with open(AML_file) as f:\n",
    "        f.readline()\n",
    "        for line in f:\n",
    "            l = line.strip().split('\\t')\n",
    "\n",
    "            pmid, sent_id, sentence, rich_sent, snps, alter,\\\n",
    "            gene, events, trigger, include_even, event_genes,\\\n",
    "            go, hpo, mesh, un_norm_bp, basic_score = line.strip().split('\\t')\n",
    "            \n",
    "            if selected_gene:\n",
    "                if gene != selected_gene:\n",
    "                    continue\n",
    "        \n",
    "\n",
    "            l_go_set = {go_term for go_term in go.split('; ')}\n",
    "            l_hpo_set = {hpo_term for hpo_term in hpo.split('; ')}\n",
    "            l_mesh_set = {mesh_term for mesh_term in mesh.split('; ')}\n",
    "            \n",
    "            # re-assignment snp and alteration \n",
    "            snp_set = set()\n",
    "            alter_set = set()\n",
    "            for snp in snps.split('; '):\n",
    "                if snp == '-':\n",
    "                    continue\n",
    "                snp_set.add(snp)\n",
    "            \n",
    "            for alt in alter.split('; '):\n",
    "                if alt == '-':\n",
    "                    continue\n",
    "                    \n",
    "                if alt.startswith('g.') or alt.startswith('c.'):\n",
    "                    snp_set.add(alt)\n",
    "                else:\n",
    "                    alter_set.add(alt)\n",
    "            \n",
    "            # get ancestor for go and hpo\n",
    "            go_ancester_set = set()\n",
    "            hpo_ancester_set = set()\n",
    "            for go_term in go.split('; '):\n",
    "                if go_term == '-':\n",
    "                    continue\n",
    "                go_id = go_term.split('-')[0]\n",
    "                if go_to_ancestor.get(go_id):\n",
    "                    go_ancester_set.add(go_to_ancestor[go_id])\n",
    "            go_ancester = '; '.join(go_ancester_set) if go_ancester_set else '-'    \n",
    "            \n",
    "            for hpo_term in hpo.split('; '):\n",
    "                if hpo_term == '-':\n",
    "                    continue\n",
    "                hpo_id = hpo_term.split('-')[0]\n",
    "                if hpo_to_ancestor.get(hpo_id):\n",
    "                    hpo_ancester_set.add(hpo_to_ancestor[hpo_id])\n",
    "            hpo_ancester = '; '.join(hpo_ancester_set) if hpo_ancester_set else '-'\n",
    "            \n",
    "            \n",
    "            #print(go_ancester)\n",
    "            \n",
    "            # start to process data for plot\n",
    "            if use_ancestor:\n",
    "                plot_go_set = set(go_ancester.split('; '))\n",
    "                plot_hpo_set = set(hpo_ancester.split('; '))\n",
    "            else:\n",
    "                plot_go_set = set(go.split('; '))\n",
    "                plot_hpo_set = set(hpo.split('; '))\n",
    "            \n",
    "            node_set.add(gene)\n",
    "            \n",
    "            \n",
    "            bp_set = plot_go_set | plot_hpo_set\n",
    "\n",
    "\n",
    "            if '-' in bp_set:\n",
    "                bp_set.remove('-')\n",
    "            if '' in bp_set:\n",
    "                bp_set.remove('')\n",
    "\n",
    "            node_set.update(bp_set)\n",
    "            \n",
    "            #print(snp_set)\n",
    "            #print(alter_set)\n",
    "            #print(bp_set)\n",
    "            \n",
    "\n",
    "            # fixme: normalized snps\n",
    "            if snp_set:\n",
    "                for snp in snp_set:\n",
    "                    snp = snp.strip()\n",
    "\n",
    "                    # fixme: some snp format is wrong in db file\n",
    "                    #if not snp.startswith('rs') or snp == 'None' \\\n",
    "                        #or not snp.startswith('p.') or not snp.startswith('c.'):\n",
    "                        #continue\n",
    "                    node_set.add('SNP')\n",
    "                    node_set.add(snp)\n",
    "\n",
    "                    # sub_var -- bp\n",
    "                    for bp in bp_set:\n",
    "                        gene_var_count[ (gene, 'SNP') ] += 1\n",
    "                        var_sub_var_count[ ('SNP', snp) ] += 1\n",
    "\n",
    "                        sub_var_bp_count[(snp, bp)] += 1\n",
    "\n",
    "            if alter_set :\n",
    "                for var in alter_set:\n",
    "                    var = var.strip()\n",
    "                    # fixme: c.175delGC in var col\n",
    "                    if len(var.split(':')) != 2:\n",
    "                        continue\n",
    "\n",
    "                    # fixme: some wrong GO HP Mesh\n",
    "                    if var.startswith(\"GO:\") or var.startswith(\"HP:\") or var.startswith('MESH'):\n",
    "                        continue\n",
    "                    var_type, sub_var_type = var.split(':')\n",
    "\n",
    "                    node_set.add(var_type)\n",
    "                    node_set.add(sub_var_type)\n",
    "\n",
    "\n",
    "                    # sub_var -- bp\n",
    "                    for bp in bp_set:\n",
    "                        gene_var_count[ (gene, var_type) ] += 1\n",
    "\n",
    "                        var_sub_var_count[ (var_type, sub_var_type) ] += 1\n",
    "\n",
    "                        sub_var_bp_count[(sub_var_type, bp)] += 1\n",
    "\n",
    "\n",
    "        # todo: only 5 sub_var_bp to show\n",
    "        nodes = [{'name': node} for node in node_set]\n",
    "\n",
    "        links = []\n",
    "        nodes = []\n",
    "        # for _key, value in (*gene_var_count.items(), *var_sub_var_count.items(), *sub_var_bp_count.items()):\n",
    "        # sub_var_bp_count = {(sub_var, bp): num for (sub_var, bp), num in sub_var_bp_count.items() if num > 5}\n",
    "        for (source, target), value in (*gene_var_count.items(), *var_sub_var_count.items(), *sub_var_bp_count.items()):\n",
    "        # for (source, target), value in (*gene_var_count.items(), *var_sub_var_count.items()):\n",
    "            # fixme: for test\n",
    "            if {'name': source} not in nodes:\n",
    "                nodes.append({'name': source} )\n",
    "            if {'name': target} not in nodes:\n",
    "                nodes.append({'name': target} )\n",
    "\n",
    "            # {\"source\": \"Total\", \"target\": \"Environment\", \"value\": 0.342284047256003}\n",
    "            links.append({\n",
    "                \"source\": source,\n",
    "                \"target\": target,\n",
    "                \"value\": value,\n",
    "            })\n",
    "    return nodes, links\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "b91b136e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_merge_db_file_add_bp_layer(merge_db_file,\n",
    "                       go_to_ancestor: dict,\n",
    "                       hpo_to_ancestor: dict,\n",
    "                       use_ancestor=False,\n",
    "                       selected_gene='',):\n",
    "    \n",
    "    \n",
    "    node_set = set()\n",
    "\n",
    "    # gene to var_type\n",
    "    gene_var_count = defaultdict(int)\n",
    "    # var_type to sub_var_type\n",
    "    var_sub_var_count = defaultdict(int)\n",
    "    # sub_var_type to bp ancestor\n",
    "    sub_var_bp_count = defaultdict(int)\n",
    "    # ancestor to bp\n",
    "    ancestor_bp_count =defaultdict(int)\n",
    "    \n",
    "       \n",
    "    with open(AML_file) as f:\n",
    "        f.readline()\n",
    "        for line in f:\n",
    "            l = line.strip().split('\\t')\n",
    "\n",
    "            pmid, sent_id, sentence, rich_sent, snps, alter,\\\n",
    "            gene, events, trigger, include_even, event_genes,\\\n",
    "            go, hpo, mesh, un_norm_bp, basic_score = line.strip().split('\\t')\n",
    "            \n",
    "            if selected_gene:\n",
    "                if gene != selected_gene:\n",
    "                    continue\n",
    "        \n",
    "\n",
    "            l_go_set = {go_term for go_term in go.split('; ')}\n",
    "            l_hpo_set = {hpo_term for hpo_term in hpo.split('; ')}\n",
    "            l_mesh_set = {mesh_term for mesh_term in mesh.split('; ')}\n",
    "            \n",
    "            # re-assignment snp and alteration \n",
    "            snp_set = set()\n",
    "            alter_set = set()\n",
    "            for snp in snps.split('; '):\n",
    "                if snp == '-':\n",
    "                    continue\n",
    "                snp_set.add(snp)\n",
    "            \n",
    "            for alt in alter.split('; '):\n",
    "                if alt == '-':\n",
    "                    continue\n",
    "                    \n",
    "                if alt.startswith('g.') or alt.startswith('c.'):\n",
    "                    snp_set.add(alt)\n",
    "                else:\n",
    "                    alter_set.add(alt)\n",
    "            \n",
    "            \n",
    "            # todo: connect bp and ancerstor term\n",
    "            bp_set = set()\n",
    "            for go_term in go.split('; '):\n",
    "                if go_term == '-':\n",
    "                    continue\n",
    "                go_id = go_term.split('-')[0]\n",
    "                if go_to_ancestor.get(go_id):\n",
    "                    go_ancestor = go_to_ancestor[go_id]\n",
    "                    bp_set.add(f'{go_term}::{go_ancestor}')\n",
    "                    \n",
    "            for hpo_term in hpo.split('; '):\n",
    "                if hpo_term == '-':\n",
    "                    continue\n",
    "                hpo_id = hpo_term.split('-')[0]\n",
    "                if hpo_to_ancestor.get(hpo_id):\n",
    "                    hpo_ancestor = hpo_to_ancestor[hpo_id]\n",
    "                    bp_set.add(f'{hpo_term}::{hpo_ancestor}')\n",
    "            \n",
    "        \n",
    "            node_set.add(gene)\n",
    "            \n",
    "            if '-' in bp_set:\n",
    "                bp_set.remove('-')\n",
    "            if '' in bp_set:\n",
    "                bp_set.remove('')\n",
    "\n",
    "            # fixme: normalized snps\n",
    "            if snp_set:\n",
    "                for snp in snp_set:\n",
    "                    snp = snp.strip()\n",
    "\n",
    "                    # fixme: some snp format is wrong in db file\n",
    "                    #if not snp.startswith('rs') or snp == 'None' \\\n",
    "                        #or not snp.startswith('p.') or not snp.startswith('c.'):\n",
    "                        #continue\n",
    "                    node_set.add('SNP')\n",
    "                    node_set.add(snp)\n",
    "\n",
    "                    # sub_var -- bp\n",
    "                    for bp_term in bp_set:\n",
    "                        bp, bp_ancestor = bp_term.split('::')\n",
    "                        \n",
    "                        gene_var_count[ (gene, 'SNP') ] += 1\n",
    "                        var_sub_var_count[ ('SNP', snp) ] += 1\n",
    "\n",
    "                        sub_var_bp_count[(snp, bp_ancestor)] += 1\n",
    "                        ancestor_bp_count[(bp_ancestor, bp)] += 1\n",
    "\n",
    "            if alter_set :\n",
    "                for var in alter_set:\n",
    "                    var = var.strip()\n",
    "                    # fixme: c.175delGC in var col\n",
    "                    if len(var.split(':')) != 2:\n",
    "                        continue\n",
    "\n",
    "                    # fixme: some wrong GO HP Mesh\n",
    "                    if var.startswith(\"GO:\") or var.startswith(\"HP:\") or var.startswith('MESH'):\n",
    "                        continue\n",
    "                    var_type, sub_var_type = var.split(':')\n",
    "\n",
    "                    node_set.add(var_type)\n",
    "                    node_set.add(sub_var_type)\n",
    "\n",
    "                    # sub_var -- bp\n",
    "                    for bp_term in bp_set:\n",
    "                        bp, bp_ancestor = bp_term.split('::')\n",
    "\n",
    "                        gene_var_count[ (gene, var_type) ] += 1\n",
    "\n",
    "                        var_sub_var_count[ (var_type, sub_var_type) ] += 1\n",
    "\n",
    "                        sub_var_bp_count[(sub_var_type, bp_ancestor)] += 1\n",
    "                        ancestor_bp_count[(bp_ancestor, bp)] += 1\n",
    "\n",
    "\n",
    "        # todo: only 5 sub_var_bp to show\n",
    "        nodes = [{'name': node} for node in node_set]\n",
    "\n",
    "        links = []\n",
    "        nodes = []\n",
    "        # for _key, value in (*gene_var_count.items(), *var_sub_var_count.items(), *sub_var_bp_count.items()):\n",
    "        # sub_var_bp_count = {(sub_var, bp): num for (sub_var, bp), num in sub_var_bp_count.items() if num > 5}\n",
    "        for (source, target), value in (*gene_var_count.items(), *var_sub_var_count.items(),\n",
    "                                        *sub_var_bp_count.items(), \n",
    "                                        #*ancestor_bp_count.items()\n",
    "                                       ):\n",
    "        # for (source, target), value in (*gene_var_count.items(), *var_sub_var_count.items()):\n",
    "            # fixme: for test\n",
    "            if {'name': source} not in nodes:\n",
    "                nodes.append({'name': source} )\n",
    "            if {'name': target} not in nodes:\n",
    "                nodes.append({'name': target} )\n",
    "\n",
    "            # {\"source\": \"Total\", \"target\": \"Environment\", \"value\": 0.342284047256003}\n",
    "            links.append({\n",
    "                \"source\": source,\n",
    "                \"target\": target,\n",
    "                \"value\": value,\n",
    "            })\n",
    "    return nodes, links, \\\n",
    "            gene_var_count, var_sub_var_count, \\\n",
    "            sub_var_bp_count, ancestor_bp_count\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "d9e4f42c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_merge_db_file_two_layer(merge_db_file,\n",
    "                       go_to_ancestor: dict,\n",
    "                       hpo_to_ancestor: dict,\n",
    "                       use_ancestor=False,\n",
    "                       selected_gene='',):\n",
    "    \n",
    "    \n",
    "    node_set = set()\n",
    "\n",
    "    # gene to var_type\n",
    "    gene_var_count = defaultdict(int)\n",
    "    # var_type to sub_var_type\n",
    "    var_sub_var_count = defaultdict(int)\n",
    "    # sub_var_type to bp\n",
    "    sub_var_bp_count = defaultdict(int)\n",
    "    \n",
    "       \n",
    "    with open(AML_file) as f:\n",
    "        f.readline()\n",
    "        for line in f:\n",
    "            l = line.strip().split('\\t')\n",
    "\n",
    "            pmid, sent_id, sentence, rich_sent, snps, alter,\\\n",
    "            gene, events, trigger, include_even, event_genes,\\\n",
    "            go, hpo, mesh, un_norm_bp, basic_score = line.strip().split('\\t')\n",
    "            \n",
    "            if selected_gene:\n",
    "                if gene != selected_gene:\n",
    "                    continue\n",
    "        \n",
    "\n",
    "            l_go_set = {go_term for go_term in go.split('; ')}\n",
    "            l_hpo_set = {hpo_term for hpo_term in hpo.split('; ')}\n",
    "            l_mesh_set = {mesh_term for mesh_term in mesh.split('; ')}\n",
    "            \n",
    "            # re-assignment snp and alteration \n",
    "            snp_set = set()\n",
    "            alter_set = set()\n",
    "            for snp in snps.split('; '):\n",
    "                if snp == '-':\n",
    "                    continue\n",
    "                snp_set.add(snp)\n",
    "            \n",
    "            for alt in alter.split('; '):\n",
    "                if alt == '-':\n",
    "                    continue\n",
    "                    \n",
    "                if alt.startswith('g.') or alt.startswith('c.'):\n",
    "                    snp_set.add(alt)\n",
    "                else:\n",
    "                    alter_set.add(alt)\n",
    "            \n",
    "            # get ancestor for go and hpo\n",
    "            go_ancester_set = set()\n",
    "            hpo_ancester_set = set()\n",
    "            for go_term in go.split('; '):\n",
    "                if go_term == '-':\n",
    "                    continue\n",
    "                go_id = go_term.split('-')[0]\n",
    "                if go_to_ancestor.get(go_id):\n",
    "                    go_ancester_set.add(go_to_ancestor[go_id])\n",
    "            go_ancester = '; '.join(go_ancester_set) if go_ancester_set else '-'    \n",
    "            \n",
    "            for hpo_term in hpo.split('; '):\n",
    "                if hpo_term == '-':\n",
    "                    continue\n",
    "                hpo_id = hpo_term.split('-')[0]\n",
    "                if hpo_to_ancestor.get(hpo_id):\n",
    "                    hpo_ancester_set.add(hpo_to_ancestor[hpo_id])\n",
    "            hpo_ancester = '; '.join(hpo_ancester_set) if hpo_ancester_set else '-'\n",
    "            \n",
    "            \n",
    "            #print(go_ancester)\n",
    "            \n",
    "            # start to process data for plot\n",
    "            if use_ancestor:\n",
    "                plot_go_set = set(go_ancester.split('; '))\n",
    "                plot_hpo_set = set(hpo_ancester.split('; '))\n",
    "            else:\n",
    "                plot_go_set = set(go.split('; '))\n",
    "                plot_hpo_set = set(hpo.split('; '))\n",
    "            \n",
    "            node_set.add(gene)\n",
    "            \n",
    "            \n",
    "            bp_set = plot_go_set | plot_hpo_set\n",
    "\n",
    "\n",
    "            if '-' in bp_set:\n",
    "                bp_set.remove('-')\n",
    "            if '' in bp_set:\n",
    "                bp_set.remove('')\n",
    "\n",
    "            node_set.update(bp_set)\n",
    "            \n",
    "            #print(snp_set)\n",
    "            #print(alter_set)\n",
    "            #print(bp_set)\n",
    "            \n",
    "\n",
    "            # fixme: normalized snps\n",
    "            if snp_set:\n",
    "                for snp in snp_set:\n",
    "                    snp = snp.strip()\n",
    "\n",
    "                    # fixme: some snp format is wrong in db file\n",
    "                    #if not snp.startswith('rs') or snp == 'None' \\\n",
    "                        #or not snp.startswith('p.') or not snp.startswith('c.'):\n",
    "                        #continue\n",
    "                    node_set.add('SNP')\n",
    "                    node_set.add(snp)\n",
    "\n",
    "                    # sub_var -- bp\n",
    "                    for bp in bp_set:\n",
    "                        gene_var_count[ (gene, 'SNPs') ] += 1\n",
    "                        #var_sub_var_count[ ('SNPs', snp) ] += 1\n",
    "\n",
    "                        sub_var_bp_count[('SNPs', bp)] += 1\n",
    "\n",
    "            if alter_set :\n",
    "                for var in alter_set:\n",
    "                    var = var.strip()\n",
    "                    # fixme: c.175delGC in var col\n",
    "                    if len(var.split(':')) != 2:\n",
    "                        continue\n",
    "\n",
    "                    # fixme: some wrong GO HP Mesh\n",
    "                    if var.startswith(\"GO:\") or var.startswith(\"HP:\") or var.startswith('MESH'):\n",
    "                        continue\n",
    "                    var_type, sub_var_type = var.split(':')\n",
    "\n",
    "                    node_set.add(var_type)\n",
    "                    node_set.add(sub_var_type)\n",
    "\n",
    "\n",
    "                    # sub_var -- bp\n",
    "                    for bp in bp_set:\n",
    "                        gene_var_count[ (gene, var_type) ] += 1\n",
    "\n",
    "                        #var_sub_var_count[ (var_type, sub_var_type) ] += 1\n",
    "\n",
    "                        sub_var_bp_count[(var_type, bp)] += 1\n",
    "\n",
    "\n",
    "        # todo: only 5 sub_var_bp to show\n",
    "        nodes = [{'name': node} for node in node_set]\n",
    "\n",
    "        links = []\n",
    "        nodes = []\n",
    "        # for _key, value in (*gene_var_count.items(), *var_sub_var_count.items(), *sub_var_bp_count.items()):\n",
    "        # sub_var_bp_count = {(sub_var, bp): num for (sub_var, bp), num in sub_var_bp_count.items() if num > 5}\n",
    "        for (source, target), value in (*gene_var_count.items(), *sub_var_bp_count.items()):\n",
    "        # for (source, target), value in (*gene_var_count.items(), *var_sub_var_count.items()):\n",
    "            # fixme: for test\n",
    "            if {'name': source} not in nodes:\n",
    "                nodes.append({'name': source} )\n",
    "            if {'name': target} not in nodes:\n",
    "                nodes.append({'name': target} )\n",
    "\n",
    "            # {\"source\": \"Total\", \"target\": \"Environment\", \"value\": 0.342284047256003}\n",
    "            links.append({\n",
    "                \"source\": source,\n",
    "                \"target\": target,\n",
    "                \"value\": value,\n",
    "            })\n",
    "    return nodes, links\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6dd5bcc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "91346262",
   "metadata": {},
   "outputs": [],
   "source": [
    "def single_type_sankey(type_to_sub_type_count: dict, \\\n",
    "                       temp_file: str, save_file: str):\n",
    "    \n",
    "    \n",
    "    nodes = []\n",
    "    links = []\n",
    "    saved_node_set = set()\n",
    "    \n",
    "    for (node, sub_node), count in type_to_sub_type_count.items():\n",
    "        \n",
    "        if node == sub_node:\n",
    "            continue\n",
    "            \n",
    "        if node == 'SNP':\n",
    "            continue\n",
    "            \n",
    "        if node not in saved_node_set:\n",
    "            nodes.append({'name': node})\n",
    "            saved_node_set.add(node)\n",
    "            \n",
    "        if sub_node not in saved_node_set:\n",
    "            nodes.append({'name': sub_node})\n",
    "            saved_node_set.add(sub_node)\n",
    "        \n",
    "        links.append({\n",
    "            \"source\": node,\n",
    "            \"target\": sub_node,\n",
    "            \"value\": count,\n",
    "        })\n",
    "        \n",
    "    with open(temp_file) as f, open(save_file, 'w') as wf:\n",
    "        doc = f.read()\n",
    "        \n",
    "        doc = doc.replace('data_str', str(nodes))\n",
    "        doc = doc.replace('link_str', str(links))\n",
    "\n",
    "        wf.write(doc)\n",
    "    print(f'{save_file} saved.')\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "7081f036",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 颜色记录 2023 07 17\n",
    "# Gene #ffa500\n",
    "# Alteration #a841da\n",
    "# BP #89ce4f\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "425a160b",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'fig_save_path' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 7\u001b[0m\n\u001b[1;32m      4\u001b[0m uas_ancestor \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m      5\u001b[0m selected_gene \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mERBB2\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 7\u001b[0m html_save_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfig_save_path\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m/ERBB2.sankey-two-layer.html\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      8\u001b[0m \u001b[38;5;66;03m#html_save_file = f'{fig_save_path}/ERBB2.sankey-ancestor-bp.html'\u001b[39;00m\n\u001b[1;32m      9\u001b[0m \n\u001b[1;32m     10\u001b[0m \u001b[38;5;66;03m#nodes, links = read_merge_db_file(AML_file, go_to_ancestor, hp_to_ancestor, uas_ancestor, selected_gene)\u001b[39;00m\n\u001b[1;32m     11\u001b[0m nodes, links \u001b[38;5;241m=\u001b[39m read_merge_db_file_two_layer(AML_file, go_to_ancestor, hp_to_ancestor, uas_ancestor, selected_gene)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'fig_save_path' is not defined"
     ]
    }
   ],
   "source": [
    "# SanKey 图保存完成\n",
    "# 替换template中对应位置\n",
    "\n",
    "uas_ancestor = True\n",
    "selected_gene = 'ERBB2'\n",
    "\n",
    "html_save_file = f'{fig_save_path}/ERBB2.sankey-two-layer.html'\n",
    "#html_save_file = f'{fig_save_path}/ERBB2.sankey-ancestor-bp.html'\n",
    "\n",
    "#nodes, links = read_merge_db_file(AML_file, go_to_ancestor, hp_to_ancestor, uas_ancestor, selected_gene)\n",
    "nodes, links = read_merge_db_file_two_layer(AML_file, go_to_ancestor, hp_to_ancestor, uas_ancestor, selected_gene)\n",
    "#nodes, links = read_merge_db_file_add_bp_layer(AML_file, go_to_ancestor, hp_to_ancestor, uas_ancestor, selected_gene)\n",
    "\n",
    "replace_template(template_html, nodes, links, html_save_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a1a55ec",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "315b9864",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 深入单基因案例\n",
    "# 没个Alteration type and GO/HPO ancestor的细分条目\n",
    "nodes, links, \\\n",
    "gene_var_count, var_sub_var_count, \\\n",
    "sub_var_bp_count, ancestor_bp_count = read_merge_db_file_add_bp_layer(AML_file, go_to_ancestor, hp_to_ancestor, uas_ancestor, selected_gene)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "84af9651",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/disk1/xzyao/CancerPNRLE/ScientificData投稿富集分析结果/ScientificData_论文绘图/单基因桑吉图绘制/var-sankey.html saved.\n"
     ]
    }
   ],
   "source": [
    "# 绘制单独的var 桑吉图\n",
    "fig_save_path = f'/mnt/disk1/xzyao/CancerPNRLE/ScientificData投稿富集分析结果/ScientificData_论文绘图/单基因桑吉图绘制'\n",
    "\n",
    "var_temp_file = f'{fig_save_path}/Sankey-template_var.html'\n",
    "var_sankey_save_file = f'{fig_save_path}/var-sankey.html'\n",
    "\n",
    "\n",
    "single_type_sankey(var_sub_var_count, var_temp_file, var_sankey_save_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "4a411866",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/disk1/xzyao/CancerPNRLE/ScientificData投稿富集分析结果/ScientificData_论文绘图/单基因桑吉图绘制/BP-sankey.html saved.\n"
     ]
    }
   ],
   "source": [
    "# 绘制单独的BP 桑吉图\n",
    "fig_save_path = f'/mnt/disk1/xzyao/CancerPNRLE/ScientificData投稿富集分析结果/ScientificData_论文绘图/单基因桑吉图绘制'\n",
    "\n",
    "bp_temp_file = f'{fig_save_path}/Sankey-template_BP.html'\n",
    "bp_sankey_save_file = f'{fig_save_path}/BP-sankey.html'\n",
    "\n",
    "\n",
    "single_type_sankey(ancestor_bp_count, bp_temp_file, bp_sankey_save_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "164a176c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def single_type_sankey(type_to_sub_type_count: dict, \\\n",
    "                       temp_file: str, save_file: str, \\\n",
    "                      one_level_color:str, second_level_color: str, \n",
    "                      display_sub_node_num:int=3):\n",
    "    \n",
    "        \n",
    "    node_count = defaultdict(int)\n",
    "    node_to_sub_node_count = defaultdict(int)\n",
    "    \n",
    "    \n",
    "    nodes = []\n",
    "    links = []\n",
    "    saved_node_set = set()\n",
    "    #print(type_to_sub_type_count)\n",
    "    for (node, sub_node), count in type_to_sub_type_count.items():\n",
    "        \n",
    "        if node == sub_node:\n",
    "            continue\n",
    "            \n",
    "        #if sub_node == 'Mutations':\n",
    "            #count = int(count/6)\n",
    "        \n",
    "        node_count[node] += count\n",
    "        node_to_sub_node_count[(node, sub_node)] += count\n",
    "\n",
    "    \n",
    "    # generate pie data in echarts\n",
    "    pie_data = []\n",
    "    for node in node_count.keys():\n",
    "        children_list = []\n",
    "        \n",
    "        sub_node_count = {sub_node: count for (_node, sub_node), count in node_to_sub_node_count.items() if _node == node}\n",
    "        \n",
    "        total_sub_node = sum(sub_node_count.values())\n",
    "        \n",
    "        # wrong value check\n",
    "        if node_count[node] != total_sub_node:\n",
    "            print(node_count[node])\n",
    "            print(sub_node_count)\n",
    "            input()\n",
    "        \n",
    "        sorted_sub_node = sorted(sub_node_count.keys(), key=lambda x: sub_node_count[x], reverse=True)\n",
    "        \n",
    "        display_count = 0\n",
    "        \n",
    "        for idx, sub_node in enumerate(sorted_sub_node):\n",
    "            sub_node_value = sub_node_count[sub_node]\n",
    "            \n",
    "            if idx >= display_sub_node_num:\n",
    "                break\n",
    "            \n",
    "            children_list.append({\n",
    "                \"name\": sub_node,\n",
    "                \"value\": sub_node_value,\n",
    "                \"itemStyle\": {\n",
    "                    \"opacity\": 0.8,\n",
    "                    \"color\": second_level_color\n",
    "                }\n",
    "            })\n",
    "            display_count += sub_node_value\n",
    "            \n",
    "            #print(children_list)\n",
    "\n",
    "        \n",
    "        other_count = total_sub_node - display_count\n",
    "        \n",
    "        if other_count > 0:\n",
    "            children_list.append({\n",
    "                \"name\": '...',\n",
    "                \"value\": other_count,\n",
    "                \"itemStyle\": {\n",
    "                \"opacity\": 0.8,\n",
    "                \"color\": second_level_color,\n",
    "                },\n",
    "                })\n",
    "        \n",
    "        pie_data.append({\n",
    "            \"name\": node,\n",
    "            \"itemStyle\": {\n",
    "                \"opacity\": 0.8,\n",
    "                \"color\": one_level_color,\n",
    "            },\n",
    "            \"value\": display_count + other_count,\n",
    "            \"children\": children_list\n",
    "            \n",
    "        })\n",
    "        \n",
    "        \n",
    "    \n",
    "    # repalce the ecahrts pie template\n",
    "    with open(temp_file) as f, open(save_file, 'w') as wf:\n",
    "        doc = f.read()\n",
    "        \n",
    "        doc = doc.replace('data_str', str(pie_data))\n",
    "\n",
    "        wf.write(doc)\n",
    "    print(f'{save_file} saved.')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "75d0ad54",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/disk1/xzyao/CancerPNRLE/ScientificData投稿富集分析结果/ScientificData_论文绘图/单基因桑吉图绘制/Alt-Pie.html saved.\n"
     ]
    }
   ],
   "source": [
    "# 绘制单独的Var饼状图\n",
    "fig_save_path = f'/mnt/disk1/xzyao/CancerPNRLE/ScientificData投稿富集分析结果/ScientificData_论文绘图/单基因桑吉图绘制'\n",
    "\n",
    "var_temp_file = f'{fig_save_path}/Pie-template.html'\n",
    "var_sankey_save_file = f'{fig_save_path}/Alt-Pie.html'\n",
    "\n",
    "display_num = 10\n",
    "\n",
    "single_type_sankey(var_sub_var_count, var_temp_file, var_sankey_save_file,\n",
    "                  '#a841da', '#b965e1', display_num)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "206ec9e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/disk1/xzyao/CancerPNRLE/ScientificData投稿富集分析结果/ScientificData_论文绘图/单基因桑吉图绘制/BP-Pie.html saved.\n"
     ]
    }
   ],
   "source": [
    "# 绘制单独的BP饼状图\n",
    "fig_save_path = f'/mnt/disk1/xzyao/CancerPNRLE/ScientificData投稿富集分析结果/ScientificData_论文绘图/单基因桑吉图绘制'\n",
    "\n",
    "bp_temp_file = f'{fig_save_path}/Pie-template.html'\n",
    "bp_sankey_save_file = f'{fig_save_path}/BP-Pie.html'\n",
    "\n",
    "display_num = 3\n",
    "\n",
    "new_ancestor_bp_count = {(source.replace('-', ' '), target.replace('-', ' ')): count for (source,target), count in ancestor_bp_count.items()}\n",
    "\n",
    "\n",
    "single_type_sankey(new_ancestor_bp_count, bp_temp_file, bp_sankey_save_file,\n",
    "                  '#89ce4f', '#a9db7f', display_num)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "a04f6bf3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys([('HP:0001871 Abnormality of blood and blood forming tissues', 'HP:0004808 Acute myeloid leukemia'), ('HP:0001871 Abnormality of blood and blood forming tissues', 'HP:0002863 Myelodysplasia'), ('GO:0065007 biological regulation', 'GO:0023052 signaling'), ('GO:0009987 cellular process', 'GO:0006915 apoptotic process'), ('GO:0009987 cellular process', 'GO:0030154 cell differentiation'), ('HP:0002664 Neoplasm', 'HP:0003002 Breast carcinoma'), ('HP:0002664 Neoplasm', 'HP:0002664 Neoplasm'), ('GO:0009987 cellular process', 'GO:0038084 vascular endothelial growth factor signaling pathway'), ('GO:0008152 metabolic process', 'GO:0009056 catabolic process'), ('GO:0003824 catalytic activity', 'GO:0016791 phosphatase activity'), ('HP:0001871 Abnormality of blood and blood forming tissues', 'HP:0001909 Leukemia'), ('HP:0002086 Abnormality of the respiratory system', 'HP:0030358 Non small cell lung carcinoma'), ('GO:0003824 catalytic activity', 'GO:0005006 epidermal growth factor receptor activity'), ('HP:0002664 Neoplasm', 'HP:0030731 Carcinoma'), ('HP:0001871 Abnormality of blood and blood forming tissues', 'HP:0002665 Lymphoma'), ('HP:0001871 Abnormality of blood and blood forming tissues', 'HP:0005523 Lymphoproliferative disorder'), ('GO:0005488 binding', 'GO:0003823 antigen binding'), ('HP:0002664 Neoplasm', 'HP:0002861 Melanoma'), ('GO:0008152 metabolic process', 'GO:0009058 biosynthetic process'), ('HP:0001574 Abnormality of the integument', 'HP:0002671 Basal cell carcinoma'), ('GO:0032501 multicellular organismal process', 'GO:0046960 sensitization'), ('GO:0003824 catalytic activity', 'GO:0016303 1 phosphatidylinositol 3 kinase activity'), ('HP:0002664 Neoplasm', 'HP:0012254 Ewing sarcoma'), ('HP:0002664 Neoplasm', 'HP:0002894 Neoplasm of the pancreas'), ('GO:0003824 catalytic activity', 'GO:0004707 MAP kinase activity'), ('GO:0008152 metabolic process', 'GO:0006351 DNA templated transcription'), ('HP:0000707 Abnormality of the nervous system', 'HP:0012174 Glioblastoma multiforme'), ('HP:0002086 Abnormality of the respiratory system', 'HP:0030359 Squamous cell lung carcinoma'), ('HP:0001871 Abnormality of blood and blood forming tissues', 'HP:0005506 Chronic myelogenous leukemia'), ('HP:0002086 Abnormality of the respiratory system', 'HP:0100526 Neoplasm of the lung'), ('HP:0000119 Abnormality of the genitourinary system', 'HP:0005584 Renal cell carcinoma'), ('HP:0002086 Abnormality of the respiratory system', 'HP:0030078 Lung adenocarcinoma'), ('GO:0009987 cellular process', 'GO:0007165 signal transduction'), ('GO:0009987 cellular process', 'GO:0016236 macroautophagy'), ('GO:0009987 cellular process', 'GO:0008283 cell population proliferation'), ('HP:0001871 Abnormality of blood and blood forming tissues', 'HP:0006775 Multiple myeloma'), ('HP:0001574 Abnormality of the integument', 'HP:0002860 Squamous cell carcinoma'), ('GO:0003824 catalytic activity', 'GO:0004697 protein kinase C activity'), ('GO:0005488 binding', 'GO:0005102 signaling receptor binding'), ('HP:0000707 Abnormality of the nervous system', 'HP:0009733 Glioma'), ('HP:0002664 Neoplasm', 'HP:0002669 Osteosarcoma'), ('GO:0009405 pathogenesis', 'GO:0009405 pathogenesis'), ('HP:0002664 Neoplasm', 'HP:0100013 Neoplasm of the breast'), ('GO:0005488 binding', 'GO:0005130 granulocyte colony stimulating factor receptor binding'), ('GO:0005488 binding', 'GO:0005129 granulocyte macrophage colony stimulating factor receptor binding'), ('HP:0001871 Abnormality of blood and blood forming tissues', 'HP:0002488 Acute leukemia'), ('HP:0000707 Abnormality of the nervous system', 'HP:0003006 Neuroblastoma'), ('HP:0025031 Abnormality of the digestive system', 'HP:0030153 Cholangiocarcinoma'), ('GO:0009987 cellular process', 'GO:0001788 antibody dependent cellular cytotoxicity'), ('HP:0001871 Abnormality of blood and blood forming tissues', 'HP:0005547 Myeloproliferative disorder'), ('GO:0005488 binding', 'GO:0005488 binding'), ('GO:0009987 cellular process', 'GO:0097194 execution phase of apoptosis'), ('HP:0000119 Abnormality of the genitourinary system', 'HP:0012125 Prostate cancer'), ('HP:0025031 Abnormality of the digestive system', 'HP:0003003 Colon cancer'), ('GO:0009987 cellular process', 'GO:0038128 ERBB2 signaling pathway'), ('HP:0025031 Abnormality of the digestive system', 'HP:0012126 Stomach cancer'), ('HP:0000119 Abnormality of the genitourinary system', 'HP:0006740 Transitional cell carcinoma of the bladder'), ('HP:0002664 Neoplasm', 'HP:0006725 Pancreatic adenocarcinoma'), ('GO:0003824 catalytic activity', 'GO:0003904 deoxyribodipyrimidine photo lyase activity'), ('GO:0008152 metabolic process', 'GO:0006306 DNA methylation'), ('HP:0002664 Neoplasm', 'HP:0012034 Liposarcoma'), ('GO:0005488 binding', 'GO:0005134 interleukin 2 receptor binding'), ('GO:0008014 calcium dependent cell adhesion molecule activity', 'GO:0008014 calcium dependent cell adhesion molecule activity'), ('GO:0003824 catalytic activity', 'GO:0005020 stem cell factor receptor activity'), ('GO:0008152 metabolic process', 'GO:0016311 dephosphorylation'), ('GO:0003824 catalytic activity', 'GO:0008177 succinate dehydrogenase (ubiquinone) activity'), ('HP:0025031 Abnormality of the digestive system', 'HP:0100723 Gastrointestinal stroma tumor'), ('HP:0002664 Neoplasm', 'HP:0012288 Neoplasm of head and neck'), ('HP:0000818 Abnormality of the endocrine system', 'HP:0002895 Papillary thyroid carcinoma'), ('GO:0040007 growth', 'GO:0040007 growth'), ('GO:0008152 metabolic process', 'GO:0010467 gene expression'), ('GO:0008152 metabolic process', 'GO:0032259 methylation'), ('GO:0008152 metabolic process', 'GO:0006412 translation'), ('GO:0008152 metabolic process', 'GO:0016310 phosphorylation'), ('GO:0032502 developmental process', 'GO:0007399 nervous system development'), ('GO:0032501 multicellular organismal process', 'GO:0007610 behavior'), ('GO:0003824 catalytic activity', 'GO:0008168 methyltransferase activity'), ('GO:0008152 metabolic process', 'GO:0006119 oxidative phosphorylation'), ('GO:0009987 cellular process', 'GO:0007049 cell cycle'), ('GO:0009987 cellular process', 'GO:0001906 cell killing'), ('GO:0050896 response to stimulus', 'GO:0006298 mismatch repair'), ('GO:0044848 biological phase', 'GO:0051320 S phase'), ('GO:0009987 cellular process', 'GO:0048242 epinephrine secretion'), ('GO:0008152 metabolic process', 'GO:0035825 homologous recombination'), ('GO:0009987 cellular process', 'GO:0016477 cell migration'), ('GO:0050896 response to stimulus', 'GO:0043627 response to estrogen'), ('GO:0050896 response to stimulus', 'GO:0032570 response to progesterone'), ('GO:0032501 multicellular organismal process', 'GO:0007613 memory'), ('GO:0005488 binding', 'GO:0003677 DNA binding'), ('GO:0009987 cellular process', 'GO:0035329 hippo signaling'), ('GO:0008152 metabolic process', 'GO:0034968 histone lysine methylation'), ('GO:0002376 immune system process', 'GO:0042119 neutrophil activation'), ('GO:0003824 catalytic activity', 'GO:0016301 kinase activity'), ('GO:0009987 cellular process', 'GO:0016049 cell growth'), ('GO:0005488 binding', 'GO:0035198 miRNA binding'), ('GO:0044848 biological phase', 'GO:0000279 M phase'), ('GO:0009987 cellular process', 'GO:0001837 epithelial to mesenchymal transition'), ('GO:0051179 localization', 'GO:0046903 secretion'), ('GO:0009987 cellular process', 'GO:0038127 ERBB signaling pathway'), ('GO:0009987 cellular process', 'GO:0008219 cell death'), ('GO:0032502 developmental process', 'GO:0001946 lymphangiogenesis'), ('GO:0009987 cellular process', 'GO:0012501 programmed cell death'), ('GO:0009987 cellular process', 'GO:0097278 complement dependent cytotoxicity'), ('GO:0009987 cellular process', 'GO:0006909 phagocytosis'), ('GO:0002376 immune system process', 'GO:0042110 T cell activation'), ('GO:0065007 biological regulation', 'GO:0031648 protein destabilization'), ('GO:0009987 cellular process', 'GO:0006914 autophagy'), ('GO:0005488 binding', 'GO:0002039 p53 binding'), ('GO:0051179 localization', 'GO:0051179 localization'), ('GO:0050896 response to stimulus', 'GO:0006302 double strand break repair'), ('GO:0002376 immune system process', 'GO:0006955 immune response'), ('GO:0002376 immune system process', 'GO:0002524 hypersensitivity'), ('GO:0050896 response to stimulus', 'GO:0009725 response to hormone'), ('GO:0032502 developmental process', 'GO:0001525 angiogenesis'), ('GO:0005488 binding', 'GO:0005524 ATP binding'), ('GO:0050896 response to stimulus', 'GO:0036274 response to lapatinib'), ('GO:0008152 metabolic process', 'GO:0070085 glycosylation'), ('GO:0044848 biological phase', 'GO:0042703 menstruation'), ('GO:0005488 binding', 'GO:0000166 nucleotide binding'), ('GO:0009987 cellular process', 'GO:0007155 cell adhesion'), ('GO:0009987 cellular process', 'GO:0072537 fibroblast activation'), ('GO:0009987 cellular process', 'GO:0097707 ferroptosis'), ('GO:0008152 metabolic process', 'GO:0006468 protein phosphorylation'), ('GO:0003824 catalytic activity', 'GO:0003824 catalytic activity'), ('GO:0005488 binding', 'GO:0003729 mRNA binding'), ('GO:0005488 binding', 'GO:0003723 RNA binding'), ('GO:0032501 multicellular organismal process', 'GO:0050817 coagulation')])"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_ancestor_bp_count.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "efc2f506",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HP:0004808 Acute myeloid leukemia 10\n",
      "HP:0002863 Myelodysplasia 4\n",
      "HP:0001909 Leukemia 2\n",
      "HP:0002665 Lymphoma 3\n",
      "HP:0005523 Lymphoproliferative disorder 3\n",
      "HP:0005506 Chronic myelogenous leukemia 4\n",
      "HP:0006775 Multiple myeloma 1\n",
      "HP:0002488 Acute leukemia 2\n",
      "HP:0005547 Myeloproliferative disorder 2\n"
     ]
    }
   ],
   "source": [
    "for (source, target), value in new_ancestor_bp_count.items():\n",
    "    if source == 'HP:0001871 Abnormality of blood and blood forming tissues':\n",
    "        print(f'{target} {value}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6afdf1d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0af022e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa003d01",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "d0da671d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(int,\n",
       "            {('copy number variation', 'deletion'): 24,\n",
       "             ('expression changes', 'overexpression'): 62,\n",
       "             ('expression changes', 'underexpression'): 101,\n",
       "             ('copy number variation', 'amplification'): 62,\n",
       "             ('point mutations', 'Mutations'): 258,\n",
       "             ('expression changes', 'inactivation'): 12,\n",
       "             ('copy number variation', 'fusion'): 19,\n",
       "             ('expression changes', 'expression'): 26,\n",
       "             ('epigenic marks', 'methylation'): 6,\n",
       "             ('epigenic marks', 'phosphorylation'): 9,\n",
       "             ('copy number variation', 'knockout'): 30,\n",
       "             ('SNP', 'rs113488022'): 9,\n",
       "             ('epigenic marks', 'epigenetics'): 13,\n",
       "             ('expression changes', 'dysregulation'): 20,\n",
       "             ('point mutations', 'insertion'): 7,\n",
       "             ('SNP', 'p.A330L'): 1,\n",
       "             ('SNP', 'p.I332E'): 1,\n",
       "             ('point mutations', 'substitution'): 1,\n",
       "             ('SNP', 'rs1468390238'): 1,\n",
       "             ('SNP', 'p.S6K'): 2,\n",
       "             ('SNP', 'p.S4C'): 1,\n",
       "             ('SNP', 'p.S4E'): 1,\n",
       "             ('SNP', 'c.2A>A'): 1,\n",
       "             ('SNP', 'p.L234F'): 1,\n",
       "             ('SNP', 'p.S239C'): 2,\n",
       "             ('SNP', 'rs121913227'): 1,\n",
       "             ('point mutations', 'Structural variation'): 1,\n",
       "             ('SNP', 'c.6A>C'): 2,\n",
       "             ('SNP', 'p.S727A'): 2,\n",
       "             ('SNP', 'rs121434592'): 1,\n",
       "             ('SNP', 'rs768227005'): 1,\n",
       "             ('SNP', 'p.S727S'): 2,\n",
       "             ('SNP', 'rs1178938625'): 2,\n",
       "             ('SNP', 'p.P396L'): 1,\n",
       "             ('SNP', 'p.L235V'): 2,\n",
       "             ('SNP', 'p.Y300L'): 1,\n",
       "             ('SNP', 'p.R292P'): 1,\n",
       "             ('SNP', 'p.F243L'): 1,\n",
       "             ('SNP', 'rs584298'): 2,\n",
       "             ('SNP', 'rs689187'): 2,\n",
       "             ('SNP', 'p.I767M'): 1,\n",
       "             ('SNP', 'rs121913468'): 1,\n",
       "             ('SNP', 'rs121913471'): 1,\n",
       "             ('epigenic marks', 'silencing'): 4,\n",
       "             ('SNP', 'p.Y187L'): 1,\n",
       "             ('SNP', 'p.E189R'): 1,\n",
       "             ('SNP', 'rs121434569'): 7,\n",
       "             ('SNP', 'rs1057519861'): 2,\n",
       "             ('SNP', 'rs121434568'): 4,\n",
       "             ('SNP', 'p.S37F'): 6,\n",
       "             ('SNP', 'p.K122N'): 6,\n",
       "             ('SNP', 'rs1237526110'): 6,\n",
       "             ('SNP', 'p.F293I'): 6,\n",
       "             ('SNP', 'rs121913529'): 7,\n",
       "             ('SNP', 'rs1482897149'): 9,\n",
       "             ('SNP', 'rs3746083'): 1,\n",
       "             ('copy number variation', 'ablation'): 3,\n",
       "             ('SNP', 'p.K334A'): 1,\n",
       "             ('SNP', 'p.E333A'): 1,\n",
       "             ('SNP', 'p.S298A'): 1,\n",
       "             ('SNP', 'rs1057519826'): 1,\n",
       "             ('SNP', 'rs919891568'): 2,\n",
       "             ('SNP', 'rs761829832'): 2,\n",
       "             ('SNP', 'rs1057519919'): 2,\n",
       "             ('SNP', 'rs121913279'): 2,\n",
       "             ('SNP', 'rs1233428419'): 2,\n",
       "             ('SNP', 'rs11554273'): 2,\n",
       "             ('SNP', 'rs397516792'): 2,\n",
       "             ('SNP', 'rs77336782'): 2,\n",
       "             ('SNP', 'rs1057519860'): 1,\n",
       "             ('epigenic marks', 'acetylation'): 2,\n",
       "             ('SNP', 'p.K27M'): 2,\n",
       "             ('SNP', 'p.S442C'): 1,\n",
       "             ('SNP', 'p.E234F'): 1,\n",
       "             ('SNP', 'p.L792H'): 1,\n",
       "             ('SNP', 'rs201992086'): 1,\n",
       "             ('SNP', 'rs942512399'): 1,\n",
       "             ('SNP', 'p.G4S'): 2,\n",
       "             ('SNP', 'p.P13K'): 1,\n",
       "             ('SNP', 'c.1GATA>GATA'): 6,\n",
       "             ('SNP', 'p.T145D'): 1,\n",
       "             ('point mutations', 'splice variant'): 1,\n",
       "             ('SNP', 'p.D9T'): 1})"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_sub_var_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9252390f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "299bd03a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('SNP', 'c.1GATA>GATA'): 6\n",
      "('SNP', 'c.2A>A'): 1\n",
      "('SNP', 'c.6A>C'): 2\n",
      "('SNP', 'p.A330L'): 1\n",
      "('SNP', 'p.D9T'): 1\n",
      "('SNP', 'p.E189R'): 1\n",
      "('SNP', 'p.E234F'): 1\n",
      "('SNP', 'p.E333A'): 1\n",
      "('SNP', 'p.F243L'): 1\n",
      "('SNP', 'p.F293I'): 6\n",
      "('SNP', 'p.G4S'): 2\n",
      "('SNP', 'p.I332E'): 1\n",
      "('SNP', 'p.I767M'): 1\n",
      "('SNP', 'p.K122N'): 6\n",
      "('SNP', 'p.K27M'): 2\n",
      "('SNP', 'p.K334A'): 1\n",
      "('SNP', 'p.L234F'): 1\n",
      "('SNP', 'p.L235V'): 2\n",
      "('SNP', 'p.L792H'): 1\n",
      "('SNP', 'p.P13K'): 1\n",
      "('SNP', 'p.P396L'): 1\n",
      "('SNP', 'p.R292P'): 1\n",
      "('SNP', 'p.S239C'): 2\n",
      "('SNP', 'p.S298A'): 1\n",
      "('SNP', 'p.S37F'): 6\n",
      "('SNP', 'p.S442C'): 1\n",
      "('SNP', 'p.S4C'): 1\n",
      "('SNP', 'p.S4E'): 1\n",
      "('SNP', 'p.S6K'): 2\n",
      "('SNP', 'p.S727A'): 2\n",
      "('SNP', 'p.S727S'): 2\n",
      "('SNP', 'p.T145D'): 1\n",
      "('SNP', 'p.Y187L'): 1\n",
      "('SNP', 'p.Y300L'): 1\n",
      "('SNP', 'rs1057519826'): 1\n",
      "('SNP', 'rs1057519860'): 1\n",
      "('SNP', 'rs1057519861'): 2\n",
      "('SNP', 'rs1057519919'): 2\n",
      "('SNP', 'rs113488022'): 9\n",
      "('SNP', 'rs11554273'): 2\n",
      "('SNP', 'rs1178938625'): 2\n",
      "('SNP', 'rs121434568'): 4\n",
      "('SNP', 'rs121434569'): 7\n",
      "('SNP', 'rs121434592'): 1\n",
      "('SNP', 'rs121913227'): 1\n",
      "('SNP', 'rs121913279'): 2\n",
      "('SNP', 'rs121913468'): 1\n",
      "('SNP', 'rs121913471'): 1\n",
      "('SNP', 'rs121913529'): 7\n",
      "('SNP', 'rs1233428419'): 2\n",
      "('SNP', 'rs1237526110'): 6\n",
      "('SNP', 'rs1468390238'): 1\n",
      "('SNP', 'rs1482897149'): 9\n",
      "('SNP', 'rs201992086'): 1\n",
      "('SNP', 'rs3746083'): 1\n",
      "('SNP', 'rs397516792'): 2\n",
      "('SNP', 'rs584298'): 2\n",
      "('SNP', 'rs689187'): 2\n",
      "('SNP', 'rs761829832'): 2\n",
      "('SNP', 'rs768227005'): 1\n",
      "('SNP', 'rs77336782'): 2\n",
      "('SNP', 'rs919891568'): 2\n",
      "('SNP', 'rs942512399'): 1\n",
      "('copy number variation', 'ablation'): 3\n",
      "('copy number variation', 'amplification'): 62\n",
      "('copy number variation', 'deletion'): 24\n",
      "('copy number variation', 'fusion'): 19\n",
      "('copy number variation', 'knockout'): 30\n",
      "('epigenic marks', 'acetylation'): 2\n",
      "('epigenic marks', 'epigenetics'): 13\n",
      "('epigenic marks', 'methylation'): 6\n",
      "('epigenic marks', 'phosphorylation'): 9\n",
      "('epigenic marks', 'silencing'): 4\n",
      "('expression changes', 'dysregulation'): 20\n",
      "('expression changes', 'expression'): 26\n",
      "('expression changes', 'inactivation'): 12\n",
      "('expression changes', 'overexpression'): 62\n",
      "('expression changes', 'underexpression'): 101\n",
      "('point mutations', 'Mutations'): 258\n",
      "('point mutations', 'Structural variation'): 1\n",
      "('point mutations', 'insertion'): 7\n",
      "('point mutations', 'splice variant'): 1\n",
      "('point mutations', 'substitution'): 1\n"
     ]
    }
   ],
   "source": [
    "sort_var = sorted(var_sub_var_count.keys())\n",
    "\n",
    "#print(sort_var)\n",
    "\n",
    "for var in sort_var:\n",
    "    print(f'{var}: {var_sub_var_count[var]}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "5ae46a12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('GO:0002376-immune system process', 'GO:0002524-hypersensitivity'): 1\n",
      "('GO:0002376-immune system process', 'GO:0006955-immune response'): 1\n",
      "('GO:0002376-immune system process', 'GO:0042110-T cell activation'): 1\n",
      "('GO:0002376-immune system process', 'GO:0042119-neutrophil activation'): 1\n",
      "('GO:0003824-catalytic activity', 'GO:0003824-catalytic activity'): 1\n",
      "('GO:0003824-catalytic activity', 'GO:0003904-deoxyribodipyrimidine photo-lyase activity'): 1\n",
      "('GO:0003824-catalytic activity', 'GO:0004697-protein kinase C activity'): 1\n",
      "('GO:0003824-catalytic activity', 'GO:0004707-MAP kinase activity'): 1\n",
      "('GO:0003824-catalytic activity', 'GO:0005006-epidermal growth factor receptor activity'): 25\n",
      "('GO:0003824-catalytic activity', 'GO:0005020-stem cell factor receptor activity'): 1\n",
      "('GO:0003824-catalytic activity', 'GO:0008168-methyltransferase activity'): 1\n",
      "('GO:0003824-catalytic activity', 'GO:0008177-succinate dehydrogenase (ubiquinone) activity'): 1\n",
      "('GO:0003824-catalytic activity', 'GO:0016301-kinase activity'): 8\n",
      "('GO:0003824-catalytic activity', 'GO:0016303-1-phosphatidylinositol-3-kinase activity'): 10\n",
      "('GO:0003824-catalytic activity', 'GO:0016791-phosphatase activity'): 3\n",
      "('GO:0005488-binding', 'GO:0000166-nucleotide binding'): 1\n",
      "('GO:0005488-binding', 'GO:0002039-p53 binding'): 1\n",
      "('GO:0005488-binding', 'GO:0003677-DNA binding'): 3\n",
      "('GO:0005488-binding', 'GO:0003723-RNA binding'): 1\n",
      "('GO:0005488-binding', 'GO:0003729-mRNA binding'): 1\n",
      "('GO:0005488-binding', 'GO:0003823-antigen binding'): 12\n",
      "('GO:0005488-binding', 'GO:0005102-signaling receptor binding'): 1\n",
      "('GO:0005488-binding', 'GO:0005129-granulocyte macrophage colony-stimulating factor receptor binding'): 1\n",
      "('GO:0005488-binding', 'GO:0005130-granulocyte colony-stimulating factor receptor binding'): 1\n",
      "('GO:0005488-binding', 'GO:0005134-interleukin-2 receptor binding'): 1\n",
      "('GO:0005488-binding', 'GO:0005488-binding'): 45\n",
      "('GO:0005488-binding', 'GO:0005524-ATP binding'): 2\n",
      "('GO:0005488-binding', 'GO:0035198-miRNA binding'): 3\n",
      "('GO:0008014-calcium-dependent cell adhesion molecule activity', 'GO:0008014-calcium-dependent cell adhesion molecule activity'): 1\n",
      "('GO:0008152-metabolic process', 'GO:0006119-oxidative phosphorylation'): 1\n",
      "('GO:0008152-metabolic process', 'GO:0006306-DNA methylation'): 1\n",
      "('GO:0008152-metabolic process', 'GO:0006351-DNA-templated transcription'): 1\n",
      "('GO:0008152-metabolic process', 'GO:0006412-translation'): 6\n",
      "('GO:0008152-metabolic process', 'GO:0006468-protein phosphorylation'): 2\n",
      "('GO:0008152-metabolic process', 'GO:0009056-catabolic process'): 1\n",
      "('GO:0008152-metabolic process', 'GO:0009058-biosynthetic process'): 5\n",
      "('GO:0008152-metabolic process', 'GO:0010467-gene expression'): 13\n",
      "('GO:0008152-metabolic process', 'GO:0016310-phosphorylation'): 19\n",
      "('GO:0008152-metabolic process', 'GO:0016311-dephosphorylation'): 2\n",
      "('GO:0008152-metabolic process', 'GO:0032259-methylation'): 8\n",
      "('GO:0008152-metabolic process', 'GO:0034968-histone lysine methylation'): 1\n",
      "('GO:0008152-metabolic process', 'GO:0035825-homologous recombination'): 6\n",
      "('GO:0008152-metabolic process', 'GO:0070085-glycosylation'): 1\n",
      "('GO:0009405-pathogenesis', 'GO:0009405-pathogenesis'): 3\n",
      "('GO:0009987-cellular process', 'GO:0001788-antibody-dependent cellular cytotoxicity'): 7\n",
      "('GO:0009987-cellular process', 'GO:0001837-epithelial to mesenchymal transition'): 1\n",
      "('GO:0009987-cellular process', 'GO:0001906-cell killing'): 6\n",
      "('GO:0009987-cellular process', 'GO:0006909-phagocytosis'): 5\n",
      "('GO:0009987-cellular process', 'GO:0006914-autophagy'): 2\n",
      "('GO:0009987-cellular process', 'GO:0006915-apoptotic process'): 1\n",
      "('GO:0009987-cellular process', 'GO:0007049-cell cycle'): 20\n",
      "('GO:0009987-cellular process', 'GO:0007155-cell adhesion'): 1\n",
      "('GO:0009987-cellular process', 'GO:0007165-signal transduction'): 5\n",
      "('GO:0009987-cellular process', 'GO:0008219-cell death'): 6\n",
      "('GO:0009987-cellular process', 'GO:0008283-cell population proliferation'): 2\n",
      "('GO:0009987-cellular process', 'GO:0012501-programmed cell death'): 8\n",
      "('GO:0009987-cellular process', 'GO:0016049-cell growth'): 5\n",
      "('GO:0009987-cellular process', 'GO:0016236-macroautophagy'): 1\n",
      "('GO:0009987-cellular process', 'GO:0016477-cell migration'): 2\n",
      "('GO:0009987-cellular process', 'GO:0030154-cell differentiation'): 4\n",
      "('GO:0009987-cellular process', 'GO:0035329-hippo signaling'): 1\n",
      "('GO:0009987-cellular process', 'GO:0038084-vascular endothelial growth factor signaling pathway'): 1\n",
      "('GO:0009987-cellular process', 'GO:0038127-ERBB signaling pathway'): 1\n",
      "('GO:0009987-cellular process', 'GO:0038128-ERBB2 signaling pathway'): 2\n",
      "('GO:0009987-cellular process', 'GO:0048242-epinephrine secretion'): 1\n",
      "('GO:0009987-cellular process', 'GO:0072537-fibroblast activation'): 1\n",
      "('GO:0009987-cellular process', 'GO:0097194-execution phase of apoptosis'): 3\n",
      "('GO:0009987-cellular process', 'GO:0097278-complement-dependent cytotoxicity'): 5\n",
      "('GO:0009987-cellular process', 'GO:0097707-ferroptosis'): 1\n",
      "('GO:0032501-multicellular organismal process', 'GO:0007610-behavior'): 2\n",
      "('GO:0032501-multicellular organismal process', 'GO:0007613-memory'): 1\n",
      "('GO:0032501-multicellular organismal process', 'GO:0046960-sensitization'): 1\n",
      "('GO:0032501-multicellular organismal process', 'GO:0050817-coagulation'): 1\n",
      "('GO:0032502-developmental process', 'GO:0001525-angiogenesis'): 2\n",
      "('GO:0032502-developmental process', 'GO:0001946-lymphangiogenesis'): 1\n",
      "('GO:0032502-developmental process', 'GO:0007399-nervous system development'): 1\n",
      "('GO:0040007-growth', 'GO:0040007-growth'): 84\n",
      "('GO:0044848-biological phase', 'GO:0000279-M phase'): 1\n",
      "('GO:0044848-biological phase', 'GO:0042703-menstruation'): 1\n",
      "('GO:0044848-biological phase', 'GO:0051320-S phase'): 2\n",
      "('GO:0050896-response to stimulus', 'GO:0006298-mismatch repair'): 6\n",
      "('GO:0050896-response to stimulus', 'GO:0006302-double-strand break repair'): 1\n",
      "('GO:0050896-response to stimulus', 'GO:0009725-response to hormone'): 2\n",
      "('GO:0050896-response to stimulus', 'GO:0032570-response to progesterone'): 1\n",
      "('GO:0050896-response to stimulus', 'GO:0036274-response to lapatinib'): 3\n",
      "('GO:0050896-response to stimulus', 'GO:0043627-response to estrogen'): 1\n",
      "('GO:0051179-localization', 'GO:0046903-secretion'): 1\n",
      "('GO:0051179-localization', 'GO:0051179-localization'): 3\n",
      "('GO:0065007-biological regulation', 'GO:0023052-signaling'): 91\n",
      "('GO:0065007-biological regulation', 'GO:0031648-protein destabilization'): 1\n",
      "('HP:0000119-Abnormality of the genitourinary system', 'HP:0005584-Renal cell carcinoma'): 1\n",
      "('HP:0000119-Abnormality of the genitourinary system', 'HP:0006740-Transitional cell carcinoma of the bladder'): 3\n",
      "('HP:0000119-Abnormality of the genitourinary system', 'HP:0012125-Prostate cancer'): 5\n",
      "('HP:0000707-Abnormality of the nervous system', 'HP:0003006-Neuroblastoma'): 4\n",
      "('HP:0000707-Abnormality of the nervous system', 'HP:0009733-Glioma'): 3\n",
      "('HP:0000707-Abnormality of the nervous system', 'HP:0012174-Glioblastoma multiforme'): 6\n",
      "('HP:0000818-Abnormality of the endocrine system', 'HP:0002895-Papillary thyroid carcinoma'): 1\n",
      "('HP:0001574-Abnormality of the integument', 'HP:0002671-Basal cell carcinoma'): 2\n",
      "('HP:0001574-Abnormality of the integument', 'HP:0002860-Squamous cell carcinoma'): 1\n",
      "('HP:0001871-Abnormality of blood and blood-forming tissues', 'HP:0001909-Leukemia'): 2\n",
      "('HP:0001871-Abnormality of blood and blood-forming tissues', 'HP:0002488-Acute leukemia'): 2\n",
      "('HP:0001871-Abnormality of blood and blood-forming tissues', 'HP:0002665-Lymphoma'): 3\n",
      "('HP:0001871-Abnormality of blood and blood-forming tissues', 'HP:0002863-Myelodysplasia'): 4\n",
      "('HP:0001871-Abnormality of blood and blood-forming tissues', 'HP:0004808-Acute myeloid leukemia'): 10\n",
      "('HP:0001871-Abnormality of blood and blood-forming tissues', 'HP:0005506-Chronic myelogenous leukemia'): 4\n",
      "('HP:0001871-Abnormality of blood and blood-forming tissues', 'HP:0005523-Lymphoproliferative disorder'): 3\n",
      "('HP:0001871-Abnormality of blood and blood-forming tissues', 'HP:0005547-Myeloproliferative disorder'): 2\n",
      "('HP:0001871-Abnormality of blood and blood-forming tissues', 'HP:0006775-Multiple myeloma'): 1\n",
      "('HP:0002086-Abnormality of the respiratory system', 'HP:0030078-Lung adenocarcinoma'): 4\n",
      "('HP:0002086-Abnormality of the respiratory system', 'HP:0030358-Non-small cell lung carcinoma'): 13\n",
      "('HP:0002086-Abnormality of the respiratory system', 'HP:0030359-Squamous cell lung carcinoma'): 2\n",
      "('HP:0002086-Abnormality of the respiratory system', 'HP:0100526-Neoplasm of the lung'): 6\n",
      "('HP:0002664-Neoplasm', 'HP:0002664-Neoplasm'): 79\n",
      "('HP:0002664-Neoplasm', 'HP:0002669-Osteosarcoma'): 1\n",
      "('HP:0002664-Neoplasm', 'HP:0002861-Melanoma'): 14\n",
      "('HP:0002664-Neoplasm', 'HP:0002894-Neoplasm of the pancreas'): 2\n",
      "('HP:0002664-Neoplasm', 'HP:0003002-Breast carcinoma'): 90\n",
      "('HP:0002664-Neoplasm', 'HP:0006725-Pancreatic adenocarcinoma'): 3\n",
      "('HP:0002664-Neoplasm', 'HP:0012034-Liposarcoma'): 2\n",
      "('HP:0002664-Neoplasm', 'HP:0012254-Ewing sarcoma'): 1\n",
      "('HP:0002664-Neoplasm', 'HP:0012288-Neoplasm of head and neck'): 2\n",
      "('HP:0002664-Neoplasm', 'HP:0030731-Carcinoma'): 6\n",
      "('HP:0002664-Neoplasm', 'HP:0100013-Neoplasm of the breast'): 6\n",
      "('HP:0025031-Abnormality of the digestive system', 'HP:0003003-Colon cancer'): 3\n",
      "('HP:0025031-Abnormality of the digestive system', 'HP:0012126-Stomach cancer'): 3\n",
      "('HP:0025031-Abnormality of the digestive system', 'HP:0030153-Cholangiocarcinoma'): 2\n",
      "('HP:0025031-Abnormality of the digestive system', 'HP:0100723-Gastrointestinal stroma tumor'): 4\n"
     ]
    }
   ],
   "source": [
    "sort_ancestor = sorted(ancestor_bp_count.keys())\n",
    "\n",
    "#print(sort_var)\n",
    "\n",
    "for ancestor in sort_ancestor:\n",
    "    print(f'{ancestor}: {ancestor_bp_count[ancestor]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48057da2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
